### PR TITLE
fix: added error handling

### PR DESCRIPTION
**Pull request by jimezesinachi**
# SynthQL pull request template

## Why

Renamed/updated the SynthqlError class and used it in the Express handler

## What changed

The SynthqlError class and Express handler
 
## Important

### CODE REVIEW

File: packages/backend/src/util/handlers/createExpressSynthqlHandler.ts [hunk 1]
```diff
@@ -17,47 +17,112 @@ export function createExpressSynthqlHandler<T>(
...
+                query,
+                returnLastOnly,
+            );
 
-                res.end();
+            writeQueryResultToResponse(res, queryResult, returnLastOnly);
+        } catch (e) {
+            if (e instanceof SynthqlError) {
+                // If it's a known error, write it to the response
+                writeErrorToResponse(res, e);
```
Comment (fhur) [inline packages/backend/src/util/handlers/createExpressSynthqlHandler.ts:35]: you should await here

File: packages/backend/src/util/handlers/createExpressSynthqlHandler.ts [hunk 2]
```diff
@@ -17,47 +17,112 @@ export function createExpressSynthqlHandler<T>(
...
-                    res.write('\n');
-                }
+            const queryResult = await tryExecuteQuery(
+                queryEngine,
+                query,
+                returnLastOnly,
+            );
 
-                res.end();
+            writeQueryResultToResponse(res, queryResult, returnLastOnly);
```
Comment (fhur) [inline packages/backend/src/util/handlers/createExpressSynthqlHandler.ts:31]: you should await here

**Review by fhur — CHANGES_REQUESTED** [review]
Good progress, but not quite 😉 . Two big points stand out:

1. I think you're adding to many try/catch without understanding what they're actually doing. 
2. The idea of the SynthqlError is to encapsulate the error logic there.  Also worth mentioning: [here's the right way](https://chatgpt.com/c/6a52c9d3-5cf4-490e-b7b1-69bce438ddfa) to extend a class in js/ts. I left a few notes in the PR around this, let me know if it's not clear what you have to do :) 
3. You're missing `await` in quite a few places!
4. Lots of tests missing :)

File: packages/backend/src/util/handlers/createExpressSynthqlHandler.ts [hunk 3]
```diff
@@ -17,47 +17,129 @@ export function createExpressSynthqlHandler<T>(
...
-                res.statusCode = 200;
-                res.setHeader('Content-Type', 'application/x-ndjson');
+                // Let another layer handle the error
+                throw e;
+            }
+        }
+    };
+}
+
+async function executeSynthqlRequest<T>(
```
Comment (fhur) [inline packages/backend/src/util/handlers/createExpressSynthqlHandler.ts:41]: Let's keep existing conventions: let's call T => DB

File: packages/backend/src/util/handlers/createExpressSynthqlHandler.ts [hunk 4]
```diff
@@ -17,47 +17,129 @@ export function createExpressSynthqlHandler<T>(
...
+        });
+    }
+}
+
+async function tryExecuteQuery<T>(
+    queryEngine: QueryEngine<T>,
+    query: any,
+    returnLastOnly: boolean,
+) {
+    if (returnLastOnly) {
```
Comment (fhur) [inline packages/backend/src/util/handlers/createExpressSynthqlHandler.ts:88]: isn't this whole function just `queryEngine.execute(query, {returnLastOnly})`? no need for an `if`

File: packages/backend/src/SynthqlError.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        return new SynthqlError(error, type, message);
+    }
+
+    static createResponseStreamingError({
+        error,
+        query,
+    }: {
+        error: any;
+        query: any;
```
Comment (fhur) [inline packages/backend/src/SynthqlError.ts:64]: we shouldn't use the `any` type here

File: packages/backend/src/SynthqlError.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    }): SynthqlError {
+        const type = 'ResponseStreamingError';
+
+        const lines = [
+            'Response streaming error!',
+            '',
+            'Failure while streaming back the results for the query:',
+            '',
+            query,
```
Comment (fhur) [inline packages/backend/src/SynthqlError.ts:73]: if query is an object this will serialize to `[Object]`, we should do JSON.strinfigy(query,null,2)

**Review by fhur — CHANGES_REQUESTED** [review]
Added some comments. Looks good overall. We'll probably be able to merge today after some added test cases 🙌

File: packages/backend/src/util/handlers/createExpressSynthqlHandler.ts (hunk 4 above)
Comment (jimezesinachi) [reply packages/backend/src/util/handlers/createExpressSynthqlHandler.ts:88]: It should be, but at the time I wrote this, I wasn't sure how passing a falsy value to returnLastOnly was being used inside. After looking inside yesterday, I can see why this is unnecessary. Will fix now

File: packages/backend/src/SynthqlError.ts (hunk 5 above)
Comment (fhur) [reply packages/backend/src/SynthqlError.ts:64]: for `query`

**Review by fhur — APPROVED** [review]

**Pull request merged**